devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+import sympy
+import numpy as np
+
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
+
+__all__ = ['Differentiable']
+
+
+class Differentiable(FrozenExpr):
+    """
+    This class represents Devito differentiable objects such as
+    sum of functions, product of function or FD approximation and
+    provides FD shortcuts for such expressions
+    """
+    _op_priority = 100.0
+    is_Function = True
+    is_Indexed = False
+
+    def __new__(cls, *args, **kwargs):
+        return sympy.Expr.__new__(cls, *args)
+
+    def __init__(self, expr, **kwargs):
+        from devito.finite_differences.finite_difference import generate_fd_functions
+        self.expr = expr.expr if isinstance(expr, Differentiable) else expr
+        self.indices = self._indices()
+        # Generate FD shortcuts for expression or copy from input
+        if isinstance(expr, Differentiable):
+            self.fd = expr.fd
+            # Recover the list of possible FD shortcuts
+            self.dtype = expr.dtype
+            self.space_order = expr.space_order
+            self.time_order = expr.time_order
+            self.staggered = expr.staggered
+        else:
+            # Recover the list of possible FD shortcuts
```

Comment:
You mean "determine"?

Reply:
move ``self.fd =...`` right below the comment?

Reply:
No mean recover from input expression